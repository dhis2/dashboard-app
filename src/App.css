body {
    background-color: #f4f6f8;
}

.app-wrapper * {
    font-family: Roboto;
}

.dashboard-wrapper {
    background-color: #f4f6f8;
    padding-left: 10px;
    padding-right: 10px;
}

/* plugin TODO */

table.pivot * {
    -webkit-user-select: none !important;
    -moz-user-select: none !important;
    -ms-user-select: none !important;
    -o-user-select: none !important;
    user-select: none !important;
}
