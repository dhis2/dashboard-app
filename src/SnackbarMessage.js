import React from 'react';

const LOADING_DASHBOARD = 'LOADING_DASHBOARD';
export const loadingDashboardMsg = { name: '', type: LOADING_DASHBOARD };

const SnackbarMessage = ({ message }) => {
    if (typeof message === 'object') {
        //future message types:  switch(message.type)
        return (
            <span>
                Loading <span style={{ fontWeight: 800 }}>{message.name}</span>{' '}
                dashboard
            </span>
        );
    }
    return message;
};

export default SnackbarMessage;
