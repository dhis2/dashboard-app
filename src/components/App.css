body {
    background-color: #f4f6f8;
}

.dashboard-wrapper {
    background-color: #f4f6f8;
    padding-left: 10px;
    padding-right: 10px;
}
