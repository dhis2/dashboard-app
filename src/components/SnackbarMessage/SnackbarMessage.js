import React from 'react'
import { connect } from 'react-redux'
import Snackbar from '@material-ui/core/Snackbar'
import PropTypes from 'prop-types'

import { sGetSnackbar } from '../../reducers/snackbar'
import { acCloseSnackbar } from '../../actions/snackbar'

const LOADING_DASHBOARD = 'LOADING_DASHBOARD'
export const loadingDashboardMsg = { name: '', type: LOADING_DASHBOARD }

const SnackbarMessageContent = ({ message }) => {
    if (typeof message === 'object') {
        //future message types:  switch(message.type)
        return (
            <span>
                Loading <span style={{ fontWeight: 800 }}>{message.name}</span>{' '}
                dashboard
            </span>
        )
    }
    return message
}

export const SnackbarMessage = props => {
    return (
        <Snackbar
            style={{ zIndex: 10001 }}
            open={props.snackbarOpen}
            message={<SnackbarMessageContent message={props.snackbarMessage} />}
            autoHideDuration={props.snackbarDuration}
            onClose={props.onCloseSnackbar}
        />
    )
}

const mapStateToProps = state => {
    const { message, duration, open } = sGetSnackbar(state)
    return {
        snackbarOpen: open,
        snackbarMessage: message,
        snackbarDuration: duration,
    }
}

SnackbarMessage.propTypes = {
    snackbarDuration: PropTypes.string,
    snackbarMessage: PropTypes.object,
    snackbarOpen: PropTypes.bool,
    onCloseSnackbar: PropTypes.func,
}

export default connect(mapStateToProps, {
    onCloseSnackbar: acCloseSnackbar,
})(SnackbarMessage)
