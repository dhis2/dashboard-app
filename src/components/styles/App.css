:root {
    /* item variables */
    --item-header-margin-top: 8px;
    --item-header-margin-bottom: 8px;
    --item-content-padding: 4px;

    /* print variables */
    /* for A4 landscape (297x210mm) */
    /* (29.7 /2.54) * 96 pixels/inch - 20px print margin */
    /* changing this number could have consequences for a multipage printout */
    --a4-landscape-width-px: 1102px;
}

body {
    margin: 0;
    padding: 0;
    font-family: sans-serif;
    background-color: var(--colors-grey200);
}

.app-shell-app {
    overflow: hidden !important;
    display: flex;
    flex-direction: column;
}

/* plugin TODO */

table.pivot * {
    -webkit-user-select: none !important;
    -moz-user-select: none !important;
    -ms-user-select: none !important;
    -o-user-select: none !important;
    user-select: none !important;
}

/* temporary workaround for headerbar */
.secondary-light-bg {
    background-color: #48a999;
}

@media print {
    body {
        inline-size: 100% !important;
        padding: 0 !important;
        margin: 0 !important;
        background-color: white;
    }

    .app-shell-app {
        overflow: visible !important;
    }
}
