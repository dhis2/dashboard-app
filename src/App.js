import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import i18n from 'dhis2-i18n';

import HeaderBarComponent from 'd2-ui/lib/app-header/HeaderBar';
import headerBarStore$ from 'd2-ui/lib/app-header/headerBar.store';
import withStateFrom from 'd2-ui/lib/component-helpers/withStateFrom';
import Snackbar from 'material-ui/Snackbar';

import PageContainer from './PageContainer/PageContainer';
import ControlBarContainer from './ControlBarContainer/ControlBarContainer';
import TitleBarCt from './TitleBar/TitleBar';
import ItemGridCt from './ItemGrid/ItemGrid';
import SnackbarMessage from './SnackbarMessage';

import { fromDashboards, fromUser, fromControlBar } from './actions';
import { acCloseSnackbar } from './actions/snackbar';
import { fromSnackbar } from './reducers';

import './App.css';

const HeaderBar = withStateFrom(headerBarStore$, HeaderBarComponent);

// App
class App extends Component {
    componentDidMount() {
        const { store, d2 } = this.context;
        store.dispatch(fromUser.acReceivedUser(d2.currentUser));
        store.dispatch(fromDashboards.tSetDashboards());
        store.dispatch(fromControlBar.tSetControlBarRows());
    }

    getChildContext() {
        return {
            baseUrl: this.props.baseUrl,
            i18n,
        };
    }

    onCloseSnackbar = () => {
        this.props.acCloseSnackbar();
    };

    render() {
        return (
            <div className="app-wrapper">
                <HeaderBar />
                <ControlBarContainer />
                <PageContainer>
                    <TitleBarCt />
                    <ItemGridCt />
                </PageContainer>
                <Snackbar
                    open={this.props.snackbarOpen}
                    message={
                        <SnackbarMessage message={this.props.snackbarMessage} />
                    }
                    autoHideDuration={this.props.snackbarDuration}
                    onRequestClose={this.props.onCloseSnackbar}
                />
            </div>
        );
    }
}

const mapStateToProps = state => {
    const { message, duration, open } = fromSnackbar.sGetSnackbar(state);
    return {
        snackbarOpen: open,
        snackbarMessage: message,
        snackbarDuration: duration,
    };
};

App.contextTypes = {
    d2: PropTypes.object,
    store: PropTypes.object,
};

App.childContextTypes = {
    baseUrl: PropTypes.string,
    i18n: PropTypes.object,
};

const AppCt = connect(mapStateToProps, {
    acCloseSnackbar,
})(App);

export default AppCt;
