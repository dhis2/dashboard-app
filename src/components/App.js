import React, { Component } from 'react';
import { HashRouter as Router, Route, Switch } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import i18n from '@dhis2/d2-i18n';
import { CssVariables } from '@dhis2/ui-core';

import { EDIT, VIEW, NEW } from './Dashboard/dashboardModes';
import { acReceivedUser } from '../actions/user';
import { tFetchDashboards } from '../actions/dashboards';
import { tSetControlBarRows } from '../actions/controlBar';
import { tSetDimensions } from '../actions/dimensions';
import Dashboard from './Dashboard/Dashboard';
import SnackbarMessage from './SnackbarMessage/SnackbarMessage';

import './App.css';

export class App extends Component {
    componentDidMount() {
        this.props.setCurrentUser(this.props.d2.currentUser);
        this.props.fetchDashboards();
        this.props.setControlBarRows();
        this.props.setDimensions(this.props.d2);
    }

    getChildContext() {
        return { baseUrl: this.props.baseUrl, i18n, d2: this.props.d2 };
    }

    render() {
        return (
            <>
                <CssVariables colors spacers />
                <Router>
                    <Switch>
                        <Route
                            exact
                            path="/"
                            render={props => (
                                <Dashboard {...props} mode={VIEW} />
                            )}
                        />
                        <Route
                            exact
                            path="/new"
                            render={props => (
                                <Dashboard {...props} mode={NEW} />
                            )}
                        />
                        <Route
                            exact
                            path="/:dashboardId"
                            render={props => (
                                <Dashboard {...props} mode={VIEW} />
                            )}
                        />
                        <Route
                            exact
                            path="/:dashboardId/edit"
                            render={props => (
                                <Dashboard {...props} mode={EDIT} />
                            )}
                        />
                    </Switch>
                </Router>
                <SnackbarMessage />
            </>
        );
    }
}

App.propTypes = {
    fetchDashboards: PropTypes.func.isRequired,
    setControlBarRows: PropTypes.func.isRequired,
    setCurrentUser: PropTypes.func.isRequired,
    setDimensions: PropTypes.func.isRequired,
    baseUrl: PropTypes.string,
    d2: PropTypes.object,
};

App.childContextTypes = {
    baseUrl: PropTypes.string,
    i18n: PropTypes.object,
    d2: PropTypes.object,
};

const mapDispatchToProps = dispatch => {
    return {
        setCurrentUser: currentUser => dispatch(acReceivedUser(currentUser)),
        fetchDashboards: () => dispatch(tFetchDashboards()),
        setControlBarRows: () => dispatch(tSetControlBarRows()),
        setDimensions: d2 => dispatch(tSetDimensions(d2)),
    };
};

export default connect(
    null,
    mapDispatchToProps
)(App);
