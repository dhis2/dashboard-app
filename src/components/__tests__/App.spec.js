import React from 'react';
import { shallow } from 'enzyme';
import { App } from '../App';
import SnackbarMessage from '../SnackbarMessage/SnackbarMessage';

jest.mock('@dhis2/ui/widgets/HeaderBar', () => () => (
    <div id="mockHeaderBar">mockHeaderBar</div>
));

jest.mock('../Dashboard/Dashboard', () => () => (
    <div id="mockDashboard">mockDashboard</div>
));

jest.mock('../../actions/dimensions', () => ({ tSetDimensions: () => null }));

describe('App', () => {
    let props;
    let shallowApp;
    let context;
    const app = context => {
        if (!shallowApp) {
            shallowApp = shallow(<App {...props} />, {
                context,
            });
        }
        return shallowApp;
    };

    beforeEach(() => {
        props = {
            d2: {},
            setCurrentUser: jest.fn(),
            fetchDashboards: jest.fn(),
            setControlBarRows: jest.fn(),
            setDimensions: jest.fn(),
        };
        shallowApp = undefined;
        context = {
            store: {
                dispatch: jest.fn(),
            },
        };
    });

    it('renders a SnackbarMessage', () => {
        expect(app(context).find(SnackbarMessage)).toHaveLength(1);
    });

    it('fetches the dashboards', () => {
        app(context);

        expect(props.fetchDashboards).toHaveBeenCalledTimes(1);
    });
});
