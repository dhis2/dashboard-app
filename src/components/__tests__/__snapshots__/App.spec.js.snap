// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the app with a dashboard 1`] = `
<div>
  <header
    style="height: 48px;"
  />
  <div
    class="viewdashboard"
  />
</div>
`;

exports[`renders the app with the start page 1`] = `
<div>
  <header
    style="height: 48px;"
  />
  <div>
    DashboardsBar
  </div>
  <div
    class="outer"
  >
    <div
      class="inner"
    >
      <div
        data-test="start-screen"
      >
        <div
          class="section"
        >
          <h3
            class="title"
          >
            Getting started
          </h3>
          <ul
            class="guide"
          >
            <li
              class="guideItem"
            >
              Search for and open saved dashboards from the top bar.
            </li>
            <li
              class="guideItem"
            >
              Click a dashboard chip to open it. Get back to this screen from the "More" menu.
            </li>
            <li
              class="guideItem"
            >
              Create a new dashboard with the + button.
            </li>
          </ul>
        </div>
        <div
          class="section"
        >
          <h3
            class="title"
          >
            Your most viewed dashboards
          </h3>
          <p>
            <a
              class="dashboard"
              href="#/rainbowdash"
            >
              <span
                class="icon"
              >
                <svg
                  color="#6e7a8a"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14 2a1 1 0 011 1v10a1 1 0 01-1 1H2a1 1 0 01-1-1V3a1 1 0 011-1zm0 4H2v7h12zM5.5 7A2.5 2.5 0 113 9.5h1A1.5 1.5 0 105.5 8zM11 7v5h-1V7zm2 2v3h-1V9zm1-6H2v2h12z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span>
                Rainbow Dash name
              </span>
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
